src/filtering.rs

```diff
@@ -116,9 +155,113 @@ impl SnapshotFilter {
+            || matches!(&self.filter_before, Some(before) if snapshot.time >= before.0)
+            || matches!((&self.filter_size,&snapshot.summary), (Some(size),Some(summary)) if !size.matches(summary.total_bytes_processed))
+            || matches!((&self.filter_size_added,&snapshot.summary), (Some(size),Some(summary)) if !size.matches(summary.data_added))
+        {
+            return false;
+        }
+
         snapshot.paths.matches(&self.filter_paths)
+            && (self.filter_paths_exact.is_empty()
+                || self.filter_paths_exact.contains(&snapshot.paths))
             && snapshot.tags.matches(&self.filter_tags)
+            && (self.filter_tags_exact.is_empty()
+                || self.filter_tags_exact.contains(&snapshot.tags))
             && (self.filter_hosts.is_empty() || self.filter_hosts.contains(&snapshot.hostname))
             && (self.filter_labels.is_empty() || self.filter_labels.contains(&snapshot.label))
     }
 }
+
+#[derive(Debug, Clone, Display)]
+struct AfterDate(DateTime<Local>);
+
+impl FromStr for AfterDate {
+    type Err = anyhow::Error;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let before_midnight = NaiveTime::from_hms_nano_opt(23, 59, 59, 999_999_999).unwrap();
+        let datetime = dateparser::parse_with(s, &Local, before_midnight)?;
+        Ok(Self(datetime.into()))
+    }
+}
+
+#[derive(Debug, Clone, Display)]
+struct BeforeDate(DateTime<Local>);
+
+impl FromStr for BeforeDate {
+    type Err = anyhow::Error;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let midnight = NaiveTime::from_hms_opt(0, 0, 0).unwrap();
+        let datetime = dateparser::parse_with(s, &Local, midnight)?;
+        Ok(Self(datetime.into()))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct SizeRange {
+    from: Option<ByteSize>,
+    to: Option<ByteSize>,
+}
+
+impl SizeRange {
+    fn matches(&self, size: u64) -> bool {
+        !matches!(self.from, Some(from) if size < from.0)
+            && !matches!(self.to, Some(to) if size > to.0)
+    }
+}
+
+fn parse_size(s: &str) -> Result<Option<ByteSize>, String> {
+    let s = s.trim();
+    if s.is_empty() {
+        return Ok(None);
+    }
+    Ok(Some(s.parse()?))
+}
+
+impl FromStr for SizeRange {
+    type Err = String;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let (from, to) = match s.split_once("..") {
+            Some((s1, s2)) => (parse_size(s1)?, parse_size(s2)?),
+            None => (parse_size(s)?, None),
+        };
+        Ok(Self { from, to })
+    }
+}
+
+impl Display for SizeRange {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if let Some(from) = self.from {
+            f.write_str(&from.to_string_as(true))?;
+        }
+        f.write_str("..")?;
+        if let Some(to) = self.to {
+            f.write_str(&to.to_string_as(true))?;
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
```

Comment:
I think it could be nice to also implement some tests for parsing `BeforeDate` and `AfterDate` to make sure we handle edge cases right.